cpp/dolfinx/mesh/Topology.cpp

```diff
@@ -116,62 +116,38 @@ compute_index_sharing(MPI_Comm comm, std::vector<std::int64_t>& unknown_indices)
 } // namespace
 
 //-----------------------------------------------------------------------------
-std::vector<bool> mesh::compute_interior_facets(const Topology& topology)
+std::vector<bool> mesh::compute_boundary_facets(const Topology& topology)
 {
-  // NOTE: Getting markers for owned and unowned facets requires reverse
-  // and forward scatters. If we can work only with owned facets we
-  // would need only a reverse scatter.
-
   const int tdim = topology.dim();
-  auto c = topology.connectivity(tdim - 1, tdim);
-  if (!c)
-    throw std::runtime_error("Facet-cell connectivity has not been computed");
-
-  auto map = topology.index_map(tdim - 1);
-  assert(map);
 
-  // Get number of connected cells for each ghost facet
-  std::vector<int> num_cells1(map->num_ghosts(), 0);
-  for (int f = 0; f < map->num_ghosts(); ++f)
+  auto facets = topology.index_map(tdim - 1);
+  if (!facets)
+    throw std::runtime_error("Facets have not been computed.");
+  std::set<std::int32_t> fwd_shared_facets;
+  if (facets->num_ghosts() == 0)
   {
-    num_cells1[f] = c->num_links(map->size_local() + f);
-    // TEST: For facet-based ghosting, an un-owned facet should be
-    // connected to only one facet
-    // if (num_cells1[f] > 1)
-    // {
-    //   throw std::runtime_error("!!!!!!!!!!");
-    //   std::cout << "!!! Problem with ghosting" << std::endl;
-    // }
-    // else
-    //   std::cout << "Facet as expected" << std::endl;
-    assert(num_cells1[f] == 1 or num_cells1[f] == 2);
+    fwd_shared_facets = std::set<std::int32_t>(
+        facets->forward_indices().begin(), facets->forward_indices().end());
   }
 
-  // Send my ghost data to owner, and receive data for my data from
-  // remote ghosts
-  std::vector<std::int32_t> owned;
-  map->scatter_rev(owned, num_cells1, 1, common::IndexMap::Mode::add);
-
-  // Mark owned facets that are connected to two cells
-  std::vector<int> num_cells0(map->size_local(), 0);
-  for (std::size_t f = 0; f < num_cells0.size(); ++f)
+  std::shared_ptr<const graph::AdjacencyList<std::int32_t>> fc
+      = topology.connectivity(tdim - 1, tdim);
+  if (!fc)
+    throw std::runtime_error("Facet-cell connectivity missing.");
+  std::vector<bool> _boundary_facet(facets->size_local(), false);
+  for (std::size_t f = 0; f < _boundary_facet.size(); ++f)
   {
-    assert(c->num_links(f) == 1 or c->num_links(f) == 2);
-    num_cells0[f] = (c->num_links(f) + owned[f]) > 1 ? 1 : 0;
+    if (fc->num_links(f) == 1
+        and fwd_shared_facets.find(f) == fwd_shared_facets.end())
+    {
+      _boundary_facet[f] = true;
```

Comment:
One-linear could go without the brackets?

Reply:
Not in this case because the preceding statement is over two lines.